### PR TITLE
feat: expose from_substrait in python package

### PR DESCRIPTION
**Pull request by gforsyth**
This is a first-cut at exposing the `from_substrait` function in the `duckdb` python package.

Initial attempts to call into the `substrait` functionality using `con.execute("CALL from_substrait('...')")` weren't working due to the way that Python stringifies protobufs (newlines, tabs, carriage returns all get serialized in a way that substrait / C++ don't like).

e.g.
In the substrait extension unittests:

```
CALL get_substrait('select count(exercise) as exercise from crossfit where dificulty_level <=5')                                                                                                                     
----         
\x12\x11\x1A\x0F\x1A\x0Dlessthanequal\x12\x11\x1A\x0F\x10\x01\x1A\x0Bis_not_null\x12\x09\x1A\x07\x10\x02\x1A\x03and\x12\x10\x1A\x0E\x10\x03\x1A\x0Acount_star\x1Ah\x12f\x0AZ:X\x12N"L\x12B\x0A@\x1A(\x1A&\x08\x02\x12\x12\x1A\x10\x12\x08\x12\x06\x0A\x04\x12\x02\x08\x01\x12\x04\x0A\x02(\x05\x12\x0E\x1A\x0C\x08\x01\x12\x08\x12\x06\x0A\x04\x12\x02\x08\x01"\x08\x0A\x06\x0A\x02\x08\x01\x0A\x00:\x0A\x0A\x08crossfit\x1A\x00"\x04\x0A\x02\x08\x03\x1A\x06\x12\x04\x0A\x02\x12\x00\x12\x08exercise
```

vs. the same call from Python

```
In [9]: res = con.execute("CALL get_substrait('select count(exercise) as exercise from crossfit where dificulty_level <=5')").fetchone()[0]

In [10]: res
Out[10]: b'\x12\x11\x1a\x0f\x1a\rlessthanequal\x12\x11\x1a\x0f\x10\x01\x1a\x0bis_not_null\x12\t\x1a\x07\x10\x02\x1a\x03and\x12\x10\x1a\x0e\x10\x03\x1a\ncount_star\x1ah\x12f\nZ:X\x12N"L\x12B\n@\x1a(\x1a&\x08\x02\x12\x12\x1a\x10\x12\x08\x12\x06\n\x04\x12\x02\x08\x01\x12\x04\n\x02(\x05\x12\x0e\x1a\x0c\x08\x01\x12\x08\x12\x06\n\x04\x12\x02\x08\x01"\x08\n\x06\n\x02\x08\x01\n\x00:\n\n\x08crossfit\x1a\x00"\x04\n\x02\x08\x03\x1a\x06\x12\x04\n\x02\x12\x00\x12\x08exercise'
```

there are a few differences in there, note `\x0Dlessthanequal` vs `\rlessthanequal` -- these are equivalent, but since we need to embed the bytes in a string in the current call from the python client, it goes badly:

```
In [11]: res = con.execute(f"CALL from_substrait('{res}::BLOB')")
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [11], in <cell line: 1>()
----> 1 res = con.execute(f"CALL from_substrait('{res}::BLOB')")

RuntimeError: Parser Error: syntax error at or near "\"
LINE 1: CALL from_substrait('b'\x12\x11\x1a\x0f\x1a\rlessthanequal\x12...
```

Now with `from_substrait` exposed, we can slap those bytes in and take it away:
```
In [14]: con.from_substrait(res)
Out[14]: 
---------------------
-- Expression Tree --
---------------------
from_substrait(\x12\x11\x1A\x0F\x1A\x0Dlessthanequal\x12\x11\x1A\x0F\x10\x01\x1A\x0Bis_not_null\x12\x09\x1A\x07\x10\x02\x1A\x03and\x12\x10\x1A\x0E\x10\x03\x1A\x0Acount_star\x1Ah\x12f\x0AZ:X\x12N"L\x12B\x0A@\x1A(\x1A&\x08\x02\x12\x12\x1A\x10\x12\x08\x12\x06\x0A\x04\x12\x02\x08\x01\x12\x04\x0A\x02(\x05\x12\x0E\x1A\x0C\x08\x01\x12\x08\x12\x06\x0A\x04\x12\x02\x08\x01"\x08\x0A\x06\x0A\x02\x08\x01\x0A\x00:\x0A\x0A\x08crossfit\x1A\x00"\x04\x0A\x02\x08\x03\x1A\x06\x12\x04\x0A\x02\x12\x00\x12\x08exercise)

---------------------
-- Result Columns  --
---------------------
- exercise (BIGINT)

---------------------
-- Result Preview  --
---------------------
exercise	
BIGINT	
[ Rows: 1]
2
```

I'm sure I've committed some C++ atrocities below, my apologies in advance.

### CODE REVIEW

**Comment by gforsyth** [comment]
Hey @pdet -- I've added in `get_substrait` and now the test roundtrips between `get_substrait` and `from_substrait`

Also ran `make format-fix` and committed those changes.  I installed `cmake-format` but I wonder if there's a version mismatch because errors out on my end (this is not important)

```
(duckdb_dev) gil@kaltenecker ~/githu/duckdb/duckdb python-substrait 🐚  cmake-format -v
0.6.11
(duckdb_dev) gil@kaltenecker ~/githu/duckdb/duckdb python-substrait 🐚  make format-fix
rm -rf src/amalgamation/*
python3 scripts/format.py --all --fix --noconfirm
Formatting all files
/home/gil/github.com/duckdb/duckdb
Failed to format file src/CMakeLists.txt
cmake-format src/CMakeLists.txt
WARNING add_xxx.py:253: Unexpected positional argument Python at 124:8
WARNING add_xxx.py:253: Unexpected positional argument Python at 133:8
WARNING add_xxx.py:253: Unexpected positional argument R at 147:6
WARNING add_xxx.py:253: Unexpected positional argument R at 160:6

make: *** [Makefile:220: format-fix] Error 1
```

File: tools/pythonpkg/duckdb-stubs/__init__.pyi [hunk 1]
```diff
@@ -6,11 +6,7 @@
 from typing import Any, ClassVar
 
 from typing import overload
-import pandas
-# stubgen override - unfortunately we need this for version checks
-import sys
-# stubgen override - This should probably not be exposed
-#_clean_default_connection: Any
+_clean_default_connection: PyCapsule
```
Comment (pdet) [inline tools/pythonpkg/duckdb-stubs/__init__.pyi:9]: The stubgen  makes mistakes here, so whenever ```# stubgen override``` is present it means these were manual changes that must be kept even if the script changes it, my bad for not letting you know before

File: tools/pythonpkg/duckdb-stubs/__init__.pyi [hunk 2]
```diff
@@ -143,9 +144,8 @@ class DuckDBPyResult:
     def fetchone(self) -> object: ...
 
 class token_type:
-    # stubgen override - these make mypy sad
-    #__doc__: ClassVar[str] = ...  # read-only
-    #__members__: ClassVar[dict] = ...  # read-only
+    __doc__: ClassVar[str] = ...  # read-only
+    __members__: ClassVar[dict] = ...  # read-only
```
Comment (pdet) [inline tools/pythonpkg/duckdb-stubs/__init__.pyi:148]: same here

File: tools/pythonpkg/duckdb-stubs/__init__.pyi [hunk 3]
```diff
@@ -157,9 +157,7 @@ class token_type:
     def __eq__(self, other: object) -> bool: ...
     def __getstate__(self) -> int: ...
     def __hash__(self) -> int: ...
-    # stubgen override - pybind only puts index in python >= 3.8: https://github.com/EricCousineau-TRI/pybind11/blob/54430436/include/pybind11/pybind11.h#L1789
-    if sys.version_info >= (3, 8):
-        def __index__(self) -> int: ...
+    def __index__(self) -> int: ...
```
Comment (pdet) [inline tools/pythonpkg/duckdb-stubs/__init__.pyi:160]: and here

**Review by pdet — COMMENTED** [review]
Hey @gforsyth Looks great!

Just some small reverts on the generated stubs file and it should be good to go!

**Comment by pdet** [comment]
> Hey @pdet -- I've added in `get_substrait` and now the test roundtrips between `get_substrait` and `from_substrait`
> 
> Also ran `make format-fix` and committed those changes. I installed `cmake-format` but I wonder if there's a version mismatch because errors out on my end (this is not important)
> 
> ```
> (duckdb_dev) gil@kaltenecker ~/githu/duckdb/duckdb python-substrait 🐚  cmake-format -v
> 0.6.11
> (duckdb_dev) gil@kaltenecker ~/githu/duckdb/duckdb python-substrait 🐚  make format-fix
> rm -rf src/amalgamation/*
> python3 scripts/format.py --all --fix --noconfirm
> Formatting all files
> /home/gil/github.com/duckdb/duckdb
> Failed to format file src/CMakeLists.txt
> cmake-format src/CMakeLists.txt
> WARNING add_xxx.py:253: Unexpected positional argument Python at 124:8
> WARNING add_xxx.py:253: Unexpected positional argument Python at 133:8
> WARNING add_xxx.py:253: Unexpected positional argument R at 147:6
> WARNING add_xxx.py:253: Unexpected positional argument R at 160:6
> 
> make: *** [Makefile:220: format-fix] Error 1
> ```

Hm, I've never come across this, my local cmake-format version is 0.6.13 though

**Comment by gforsyth** [comment]
All good.  Just reverted the overrides

**Comment by gforsyth** [comment]
Hey @pdet -- I believe most of the failures here are due to CI hiccups and unrealted to my changes.  The Python 3.9 Linux job is definitely failing though.

I can reproduce this locally but I'm unsure of what's going on -- if I run `make` locally, the python package that gets installed has `from_substrait` and `get_substrait` functions that work.
If I build a wheel using `python setup.py bdist_wheel`, those functions are also available but throw 

```
RuntimeError: Catalog Error: Table Function with name get_substrait does not exist!
Did you mean "summary"?
```

I've clearly missed adding something somewhere -- any thoughts?  Thanks!

**Comment by gforsyth** [comment]
:tada: CI is green!  Ready for another look @pdet !

**Review by pdet — APPROVED** [review]
I think it looks great! Thanks.

@Mytherin if you have no requests, I think this is ready to merge.

**Pull request merged**